service-base/src/main/java/org/eclipse/hono/service/credentials/CredentialsPlainPasswordHelper.java

```diff
@@ -0,0 +1,100 @@
+/*******************************************************************************
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.service.credentials;
+
+import io.vertx.core.json.JsonObject;
+import org.eclipse.hono.util.ClearTextPassword;
+import org.eclipse.hono.util.CredentialsConstants;
+
+import java.security.SecureRandom;
+
+/**
+ * Utility methods for hashing of plain text passwords for Credentials endpoint.
+ */
+public class CredentialsPlainPasswordHelper {
+
+    /**
+     * The cryptographically maximal strong salt for SHA-256 in bytes.
+     */
+    private static final int SALT_LENGTH_SHA256 = 32;
+    /**
+     * The cryptographically maximal strong salt for SHA-512 in bytes.
+     */
+    private static final int SALT_LENGTH_SHA512 = 64;
+
+    private static final String DEFAULT_HASH_FUNCTION = CredentialsConstants.HASH_FUNCTION_BCRYPT;
+
+    private static final SecureRandom SECURE_RANDOM = new SecureRandom();
+
+    private CredentialsPlainPasswordHelper() {
+        // prevent instantiation
+    }
+
+    /**
+     * Hashes a plain text password in the field @link{CredentialsConstants.FIELD_SECRETS_PWD_PLAIN} and updates the
+     * secret accordingly.
+     *
+     * @param secret The secret to be updated.
+     * @return The updated secret.
+     */
+    public static JsonObject hashPwdAndUpdateSecret(final JsonObject secret) {
+
+        if (secret.getString(CredentialsConstants.FIELD_SECRETS_PWD_PLAIN, "").isEmpty()) {
```

Comment:
what happens if a client provides an empty cleartext pwd?
```
{ "pwd-plain": "" }
```

Reply:
The client gets a HTTP 400 as for all other client errors in the Credentials REST API. We could at least log explicitly that the password is empty, which is consistent to the current state of client errors in the Credentials REST API.

Reply:
ok, so we let this case *fall through* to the service implementation which will (hopefully) detect that the *pwd-hash* field is missing and will return a 400, right?

Reply:
In my last comment I meant that we should throw an exception if the clear text password is empty.

Reply:
I understand, however, I think we can safely let the service implementation handle this problem. It needs to do that anyway, don't you think?

Reply:
So, you are saying we could leave this line as it is in the snippet above?

Reply:
I think so.